pkg/autodiscovery/providers/prometheus_services.go

```diff
@@ -65,36 +128,79 @@ func (p *PrometheusServicesConfigProvider) String() string {
 
 // Collect retrieves services from the apiserver, builds Config objects and returns them
 func (p *PrometheusServicesConfigProvider) Collect() ([]integration.Config, error) {
-	services, err := p.lister.List(labels.Everything())
+	services, err := p.api.ListServices()
 	if err != nil {
 		return nil, err
 	}
 
-	p.upToDate = true
-	return p.parseServices(services), nil
-}
-
-// IsUpToDate allows to cache configs as long as no changes are detected in the apiserver
-func (p *PrometheusServicesConfigProvider) IsUpToDate() (bool, error) {
-	return p.upToDate, nil
-}
-
-// parseServices returns a list of configurations based on the service annotations
-func (p *PrometheusServicesConfigProvider) parseServices(services []*v1.Service) []integration.Config {
 	var configs []integration.Config
 	for _, svc := range services {
 		for _, check := range p.checks {
-			configs = append(configs, check.ConfigsForService(svc)...)
+			serviceConfigs := check.ConfigsForService(svc)
+
+			if len(serviceConfigs) == 0 {
+				continue
+			}
+
+			configs = append(configs, serviceConfigs...)
+
+			if !p.collectEndpoints {
+				continue
+			}
+
+			ep, err := p.api.GetEndpoints(svc.GetNamespace(), svc.GetName())
+			if err != nil {
+				return nil, err
+			}
+
+			endpointConfigs := check.ConfigsForServiceEndpoints(svc, ep)
+
+			if len(endpointConfigs) == 0 {
+				continue
+			}
+
+			configs = append(configs, endpointConfigs...)
+
+			endpointsID := apiserver.EntityForEndpoints(ep.GetNamespace(), ep.GetName(), "")
+			p.Lock()
+			p.monitoredEndpoints[endpointsID] = true
+			p.Unlock()
+
 		}
 	}
-	return configs
+
+	p.setUpToDate(true)
+	return configs, nil
+}
+
+// setUpToDate is a thread-safe method to update the upToDate value
+func (p *PrometheusServicesConfigProvider) setUpToDate(v bool) {
+	p.Lock()
+	defer p.Unlock()
+	p.upToDate = v
+}
+
+// IsUpToDate allows to cache configs as long as no changes are detected in the apiserver
+func (p *PrometheusServicesConfigProvider) IsUpToDate() (bool, error) {
+	p.RLock()
+	defer p.RUnlock()
+	return p.upToDate, nil
 }
 
 func (p *PrometheusServicesConfigProvider) invalidate(obj interface{}) {
-	if obj != nil {
-		log.Trace("Invalidating configs on new/deleted service")
-		p.upToDate = false
+	castedObj, ok := obj.(*v1.Service)
+	if !ok {
+		log.Errorf("Expected a Service type, got: %T", obj)
+		return
+	}
+	p.Lock()
+	defer p.Unlock()
+	if p.collectEndpoints {
+		endpointsID := apiserver.EntityForEndpoints(castedObj.Namespace, castedObj.Name, "")
+		log.Tracef("Invalidating configs on new/deleted service, endpoints entity: %s", endpointsID)
+		delete(p.monitoredEndpoints, endpointsID)
 	}
+	p.upToDate = false
```

Comment:
should we use `p.setUpToDate(false)` ? why don't we add thread safe getters and setters to all the attributes we need?

Reply:
It's already under the lock, we are performing two operations here - clearing an entry in the map and updating the `upToDate` field. The mutex used here is not reentrant.

Reply:
yes saw the lock, it just feel more natural to use `setUpToDate` to set `upToDate`